src/types/websocket.events.ts

```diff
@@ -42,57 +49,149 @@ export interface WSOrderbookEventV5 {
+  category: CategoryV5;
+  orderId: string;
+  orderLinkId: string;
+  isLeverage: string;
+  blockTradeId: string;
+  symbol: string;
+  price: string;
+  qty: string;
   side: OrderSideV5;
-  placeType: string;
+  positionIdx: PositionIdx;
+  orderStatus: OrderStatusV5;
+  createType: OrderCreateTypeV5;
+  cancelType: OrderCancelTypeV5;
+  rejectReason?: OrderRejectReasonV5;
   avgPrice?: string;
   leavesQty?: string;
-  isLeverage: string;
-  cancelType: string;
+  leavesValue?: string;
   cumExecQty: string;
+  cumExecValue: string;
   cumExecFee: string;
-  smpOrderId: string;
-  takeProfit: string;
-  reduceOnly: boolean;
-  orderLinkId: string;
-  positionIdx: number;
-  tpTriggerBy: string;
-  slTriggerBy: string;
-  createdTime: string;
-  updatedTime: string;
   feeCurrency: string;
+  timeInForce: OrderTimeInForceV5;
+  orderType: OrderTypeV5;
+  stopOrderType: StopOrderTypeV5;
+  ocoTriggerType?: OCOTriggerTypeV5;
+  orderIv: string;
+  marketUnit?: 'baseCoin' | 'quoteCoin';
   triggerPrice: string;
-  category: CategoryV5;
-  cumExecValue: string;
-  blockTradeId: string;
-  leavesValue?: string;
-  slLimitPrice?: string;
-  tpLimitPrice?: string;
+  takeProfit: string;
+  stopLoss: string;
   tpslMode?: TPSLModeV5;
-  orderType: OrderTypeV5;
-  smpType: OrderSMPTypeV5;
-  closeOnTrigger: boolean;
+  tpLimitPrice?: string;
+  slLimitPrice?: string;
+  tpTriggerBy: string;
+  slTriggerBy: string;
   triggerDirection: number;
-  orderStatus: OrderStatusV5;
-  lastPriceOnCreated: string;
   triggerBy: OrderTriggerByV5;
-  stopOrderType: StopOrderTypeV5;
-  timeInForce: OrderTimeInForceV5;
-  ocoTriggerType?: OCOTriggerTypeV5;
-  rejectReason?: OrderRejectReasonV5;
+  lastPriceOnCreated: string;
+  reduceOnly: boolean;
+  closeOnTrigger: boolean;
+  placeType: string;
+  smpType: OrderSMPTypeV5;
+  smpGroup: number;
+  smpOrderId: string;
+  createdTime: string;
+  updatedTime: string;
 }
 
 export interface WSAccountOrderEventV5 {
   id: string;
-  wsKey: WsKey;
   topic: 'order';
   creationTime: number;
   data: WSAccountOrderV5[];
+  wsKey: WsKey;
+}
+
+export interface WSExecutionV5 {
+  category: CategoryV5;
+  symbol: string;
+  isLeverage: string;
+  orderId: string;
+  orderLinkId: string;
+  side: OrderSideV5;
+  orderPrice: string;
+  orderQty: string;
+  leavesQty: string;
+  createType: OrderCreateTypeV5;
+  orderType: OrderTypeV5;
+  stopOrderType: StopOrderTypeV5;
+  execFee: string;
+  execId: string;
+  execPrice: string;
+  execQty: string;
+  execType: ExecTypeV5;
+  execValue: string;
+  execTime: string;
+  isMaker: boolean;
+  feeRate: string;
+  tradeIv: string;
+  markIv: string;
+  markPrice: string;
+  indexPrice: string;
+  underlyingPrice: string;
+  blockTradeId: string;
+  closedSize: string;
+  seq: number;
+  marketUnit: string;
 }
+
+export interface WSExecutionEventV5 {
+  id: string;
+  topic: 'execution';
+  creationTime: number;
+  data: WSExecutionV5[];
+  wsKey: WsKey;
+}
```

Comment:
These top-level events all seem to have a similar shape. Minor optimisation here, but I think it's time to introduce a generic to define the shape of each top level event. This would be the generic:
```typescript
export interface WSTopicEventV5<TTopic extends string, TData> {
  id: string;
  topic: TTopic;
  creationTime: number;
  data: TData;
  wsKey: WsKey;
}
```

And then the structure of any topic event could simply be defined as such, using your WSExecutionEventV5 & WSPositionEventV5 as examples:
```typescript
export type WSExecutionEventV5 = WSTopicEventV5<'execution', WSExecutionV5[]>;
export type WSPositionEventV5 = WSTopicEventV5<'position', WSPositionV5[]>;
```

The end result is the same, but I think it's worth introducing now where the repetition starts. Let me know if you have questions!

Reply:
There's a tiny difference between the private and public streams. For example `ts` for public streams instead of `creationTime`. Would we then make a `WSPrivateTopicEventV5` and `WSPublicTopicEventV5`?

Reply:
> Would we then make a WSPrivateTopicEventV5 and WSPublicTopicEventV5

Yes, I would do exactly that.